File: backend/app/models/referral.py
from datetime import datetime, timezone
from typing import Optional
import uuid

import sqlalchemy as sa
from sqlalchemy import Column, DateTime, SmallInteger, Text
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlmodel import Field, SQLModel

from app.models.referral_state import ReferralState


class Referral(SQLModel, table=True):
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    patient_id: uuid.UUID
    from_facility_id: uuid.UUID
    destination_facility_id: uuid.UUID
    reason: str
    urgency: str

    # D2-S7: FIX -- `status` was previously declared `status: str = "INITIATED"`,
    # which SQLAlchemy bound as VARCHAR on every INSERT/UPDATE. Migration
    # d4f1c9b7a582 already converted the live `referral.status` column to a
    # real 13-state Postgres enum (`referral_status`), so every write through
    # the old `str` mapping failed with psycopg's DatatypeMismatch ("column
    # status is of type referral_status but expression is of type character
    # varying"). Fixed the same way app/models/triage_encounter.py's own
    # decision columns were fixed after their own migration (a4d72f9e1c83):
    # an explicit sa_column carrying the real SQLAlchemy type.
    # `create_type=False` is deliberate -- the Postgres type already exists
    # (created idempotently by the migration itself); SQLAlchemy must never
    # attempt to CREATE TYPE on table-create/metadata operations, which
    # would collide on every environment that already ran the migration.
    # ReferralState's members are a `str, Enum` with member NAME == member
    # VALUE for all 13 states (see app/models/referral_state.py), so
    # SQLAlchemy's default name-based enum binding already round-trips
    # correctly against the DB values written by the migration's own Phase-3
    # backfill -- no separate `values_callable` needed.
    status: ReferralState = Field(
        default=ReferralState.INITIATED,
        sa_column=Column(
            "status",
            sa.Enum(ReferralState, name="referral_status", create_type=False),
            nullable=False,
        ),
    )

    receiving_unit: Optional[str] = None
    owner: Optional[str] = None
    due_date: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    # S20: see app/models/patient.py's identical note -- migration
    # 9d5f6b3e0a71 (S10) already added these as NOT NULL DB columns;
    # the route always sets both from the authenticated actor, never
    # trusts a client-supplied value.
    created_by_user_id: Optional[uuid.UUID] = None
    org_unit_id: Optional[uuid.UUID] = None

    # ------------------------------------------------------------------
    # D2-S7: referral workflow columns added by migration d4f1c9b7a582.
    # This model was not updated to match until now -- same gap as
    # patient.py/triage_encounter.py's own earlier S20 notes (see that
    # migration's own docstring for the nullable/NOT-NULL reasoning per
    # column). The route (app/api/routes/referrals.py) is the only writer
    # of these; a client can technically submit them on POST /referrals/
    # (the whole model is the request body there, same pre-existing
    # pattern as created_by_user_id/org_unit_id above) but every one of
    # them is functionally inert at creation time since the referral is
    # always INITIATED on creation and these are all later-transition
    # fields -- fixing that (if desired) is a separate concern, out of
    # scope for this step, which is the PATCH .../status endpoint only.
    # ------------------------------------------------------------------
    # default_factory uses an explicit tz-aware UTC `now()`, not the naive
    # `datetime.utcnow` pattern `created_at` above uses -- `created_at` is
    # backed by a naive `DateTime()` column (migration 79a9d8f8db61) where
    # that's correct, but `initiated_at` is `TIMESTAMPTZ` (migration
    # d4f1c9b7a582), and the migration's own docstring flags exactly this
    # class of bug: inserting a naive datetime into a timestamptz column
    # is interpreted per the DB session's own `TimeZone` setting, not
    # assumed UTC. A tz-aware value removes that ambiguity entirely.
    initiated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column("initiated_at", DateTime(timezone=True), nullable=False),
    )
    due_at: Optional[datetime] = Field(default=None, sa_column=Column("due_at", DateTime(timezone=True)))
    breached_at: Optional[datetime] = Field(default=None, sa_column=Column("breached_at", DateTime(timezone=True)))
    breach_detected_by: Optional[str] = Field(default=None, sa_column=Column("breach_detected_by", Text))
    escalation_stage: int = Field(default=0, sa_column=Column("escalation_stage", SmallInteger, nullable=False))
    escalation_notified_at: Optional[datetime] = Field(
        default=None, sa_column=Column("escalation_notified_at", DateTime(timezone=True))
    )
    owner_user_id: Optional[uuid.UUID] = Field(default=None, sa_column=Column("owner_user_id", PGUUID(as_uuid=True)))
    arrived_at: Optional[datetime] = Field(default=None, sa_column=Column("arrived_at", DateTime(timezone=True)))
    back_referred_at: Optional[datetime] = Field(
        default=None, sa_column=Column("back_referred_at", DateTime(timezone=True))
    )
    closed_at: Optional[datetime] = Field(default=None, sa_column=Column("closed_at", DateTime(timezone=True)))
    slot_datetime: Optional[datetime] = Field(default=None, sa_column=Column("slot_datetime", DateTime(timezone=True)))
    transport_mode: Optional[str] = Field(default=None, sa_column=Column("transport_mode", Text))
    arrival_confirmed_by: Optional[uuid.UUID] = Field(
        default=None, sa_column=Column("arrival_confirmed_by", PGUUID(as_uuid=True))
    )
    arrival_scan_ref: Optional[str] = Field(default=None, sa_column=Column("arrival_scan_ref", Text))
    refusal_reason: Optional[str] = Field(default=None, sa_column=Column("refusal_reason", Text))
    loss_reason: Optional[str] = Field(default=None, sa_column=Column("loss_reason", Text))
    cancellation_reason: Optional[str] = Field(default=None, sa_column=Column("cancellation_reason", Text))
    back_referral_note: Optional[str] = Field(default=None, sa_column=Column("back_referral_note", Text))

    # Dashboard step: migration b9e4c7a2f815 added this nullable column.
    # NEVER set by any route today -- `referral` has no `client_uuid`
    # column to match an incoming sync batch record against (see that
    # migration's own docstring "THE SYNC/synced_at GAP" and
    # app/api/routes/sync.py's own module docstring). Stays NULL until a
    # future, separate migration adds `client_uuid` here.
    synced_at: Optional[datetime] = Field(default=None, sa_column=Column("synced_at", DateTime(timezone=True)))

File: frontend/lib/prototype/models/referral.dart
enum ReferralState {
  initiated,
  slotBooked,
  transportArranged,
  arrived,
  consulted,
  backReferred,
  closed,
  cancelled,
  notArrived,
  traced,
  rescheduled,
  refused,
  lost,
}

enum RefusalReason {
  cost,
  distance,
  familyDecidedAgainst,
  alreadyRecovered,
  wentPrivate,
  couldNotBeContacted,
  other,
}

const _referralStateValues = {
  ReferralState.initiated: 'INITIATED',
  ReferralState.slotBooked: 'SLOT_BOOKED',
  ReferralState.transportArranged: 'TRANSPORT_ARRANGED',
  ReferralState.arrived: 'ARRIVED',
  ReferralState.consulted: 'CONSULTED',
  ReferralState.backReferred: 'BACK_REFERRED',
  ReferralState.closed: 'CLOSED',
  ReferralState.cancelled: 'CANCELLED',
  ReferralState.notArrived: 'NOT_ARRIVED',
  ReferralState.traced: 'TRACED',
  ReferralState.rescheduled: 'RESCHEDULED',
  ReferralState.refused: 'REFUSED',
  ReferralState.lost: 'LOST',
};

const _refusalReasonValues = {
  RefusalReason.cost: 'COST',
  RefusalReason.distance: 'DISTANCE',
  RefusalReason.familyDecidedAgainst: 'FAMILY_DECIDED_AGAINST',
  RefusalReason.alreadyRecovered: 'ALREADY_RECOVERED',
  RefusalReason.wentPrivate: 'WENT_PRIVATE',
  RefusalReason.couldNotBeContacted: 'COULD_NOT_BE_CONTACTED',
  RefusalReason.other: 'OTHER',
};

String referralStateValue(ReferralState value) => _referralStateValues[value]!;

ReferralState referralStateFromValue(String value) => _referralStateValues
    .entries
    .firstWhere(
      (entry) => entry.value == value,
      orElse: () => throw FormatException('Unknown referral state: $value'),
    )
    .key;

String refusalReasonValue(RefusalReason value) => _refusalReasonValues[value]!;

class Referral {
  const Referral({
    required this.patientId,
    required this.fromFacilityId,
    required this.destinationFacilityId,
    required this.reason,
    required this.urgency,
    this.id,
    this.status = ReferralState.initiated,
    this.receivingUnit,
    this.owner,
    this.dueDate,
    this.createdAt,
    this.createdByUserId,
    this.orgUnitId,
    this.initiatedAt,
    this.dueAt,
    this.breachedAt,
    this.breachDetectedBy,
    this.escalationStage = 0,
    this.escalationNotifiedAt,
    this.ownerUserId,
    this.arrivedAt,
    this.backReferredAt,
    this.closedAt,
    this.slotDatetime,
    this.transportMode,
    this.arrivalConfirmedBy,
    this.arrivalScanRef,
    this.refusalReason,
    this.lossReason,
    this.cancellationReason,
    this.backReferralNote,
    this.syncedAt,
  });

  final String? id;
  final String patientId;
  final String fromFacilityId;
  final String destinationFacilityId;
  final String reason;
  final String urgency;
  final ReferralState status;
  final String? receivingUnit;
  final String? owner;
  final DateTime? dueDate;
  final DateTime? createdAt;
  final String? createdByUserId;
  final String? orgUnitId;
  final DateTime? initiatedAt;
  final DateTime? dueAt;
  final DateTime? breachedAt;
  final String? breachDetectedBy;
  final int escalationStage;
  final DateTime? escalationNotifiedAt;
  final String? ownerUserId;
  final DateTime? arrivedAt;
  final DateTime? backReferredAt;
  final DateTime? closedAt;
  final DateTime? slotDatetime;
  final String? transportMode;
  final String? arrivalConfirmedBy;
  final String? arrivalScanRef;
  final String? refusalReason;
  final String? lossReason;
  final String? cancellationReason;
  final String? backReferralNote;
  final DateTime? syncedAt;

  factory Referral.fromJson(Map<String, dynamic> json) => Referral(
    id: json['id'] as String?,
    patientId: json['patient_id'] as String,
    fromFacilityId: json['from_facility_id'] as String,
    destinationFacilityId: json['destination_facility_id'] as String,
    reason: json['reason'] as String,
    urgency: json['urgency'] as String,
    status: json['status'] == null
        ? ReferralState.initiated
        : referralStateFromValue(json['status'] as String),
    receivingUnit: json['receiving_unit'] as String?,
    owner: json['owner'] as String?,
    dueDate: _date(json['due_date']),
    createdAt: _date(json['created_at']),
    createdByUserId: json['created_by_user_id'] as String?,
    orgUnitId: json['org_unit_id'] as String?,
    initiatedAt: _date(json['initiated_at']),
    dueAt: _date(json['due_at']),
    breachedAt: _date(json['breached_at']),
    breachDetectedBy: json['breach_detected_by'] as String?,
    escalationStage: json['escalation_stage'] as int? ?? 0,
    escalationNotifiedAt: _date(json['escalation_notified_at']),
    ownerUserId: json['owner_user_id'] as String?,
    arrivedAt: _date(json['arrived_at']),
    backReferredAt: _date(json['back_referred_at']),
    closedAt: _date(json['closed_at']),
    slotDatetime: _date(json['slot_datetime']),
    transportMode: json['transport_mode'] as String?,
    arrivalConfirmedBy: json['arrival_confirmed_by'] as String?,
    arrivalScanRef: json['arrival_scan_ref'] as String?,
    refusalReason: json['refusal_reason'] as String?,
    lossReason: json['loss_reason'] as String?,
    cancellationReason: json['cancellation_reason'] as String?,
    backReferralNote: json['back_referral_note'] as String?,
    syncedAt: _date(json['synced_at']),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'patient_id': patientId,
    'from_facility_id': fromFacilityId,
    'destination_facility_id': destinationFacilityId,
    'reason': reason,
    'urgency': urgency,
    'status': referralStateValue(status),
    'receiving_unit': receivingUnit,
    'owner': owner,
    'due_date': dueDate?.toIso8601String(),
    'created_at': createdAt?.toIso8601String(),
    'created_by_user_id': createdByUserId,
    'org_unit_id': orgUnitId,
    'initiated_at': initiatedAt?.toIso8601String(),
    'due_at': dueAt?.toIso8601String(),
    'breached_at': breachedAt?.toIso8601String(),
    'breach_detected_by': breachDetectedBy,
    'escalation_stage': escalationStage,
    'escalation_notified_at': escalationNotifiedAt?.toIso8601String(),
    'owner_user_id': ownerUserId,
    'arrived_at': arrivedAt?.toIso8601String(),
    'back_referred_at': backReferredAt?.toIso8601String(),
    'closed_at': closedAt?.toIso8601String(),
    'slot_datetime': slotDatetime?.toIso8601String(),
    'transport_mode': transportMode,
    'arrival_confirmed_by': arrivalConfirmedBy,
    'arrival_scan_ref': arrivalScanRef,
    'refusal_reason': refusalReason,
    'loss_reason': lossReason,
    'cancellation_reason': cancellationReason,
    'back_referral_note': backReferralNote,
    'synced_at': syncedAt?.toIso8601String(),
  };
}

class ReferralStatusUpdateBody {
  const ReferralStatusUpdateBody({
    this.reason,
    this.slotDatetime,
    this.destinationOrgUnitId,
    this.transportMode,
    this.arrivalConfirmedBy,
    this.arrivalScanRef,
    this.consultedByUserId,
    this.backReferralNote,
    this.tracedByUserId,
    this.refusalReason,
    this.lossReason,
    this.cancellationReason,
  });

  final String? reason;
  final DateTime? slotDatetime;
  final String? destinationOrgUnitId;
  final String? transportMode;
  final String? arrivalConfirmedBy;
  final String? arrivalScanRef;
  final String? consultedByUserId;
  final String? backReferralNote;
  final String? tracedByUserId;
  final RefusalReason? refusalReason;
  final String? lossReason;
  final String? cancellationReason;

  Map<String, dynamic> toJson() => {
    'reason': reason,
    'slot_datetime': slotDatetime?.toIso8601String(),
    'destination_org_unit_id': destinationOrgUnitId,
    'transport_mode': transportMode,
    'arrival_confirmed_by': arrivalConfirmedBy,
    'arrival_scan_ref': arrivalScanRef,
    'consulted_by_user_id': consultedByUserId,
    'back_referral_note': backReferralNote,
    'traced_by_user_id': tracedByUserId,
    'refusal_reason': refusalReason == null
        ? null
        : refusalReasonValue(refusalReason!),
    'loss_reason': lossReason,
    'cancellation_reason': cancellationReason,
  };
}

DateTime? _date(Object? value) =>
    value == null ? null : DateTime.parse(value as String);
